post.md

```diff
@@ -0,0 +1,166 @@
+
+Here's an example of a Zod schema definition:
+
+```ts
+import { z } from "zod";
+
+const schema = z.object({
+  name: z.string().min(1, "Name cannot be blank"),
+  email: z.string().email("Invalid email address"),
+});
+```
+
+## Handling Server-Side Validation Errors
+
+When we submit our form, the server-side action validates the form data against the Zod schema. If the validation fails, the server returns an array of errors (`ZodIssue[]`). Each error object has the following structure:
+
+```json
+{
+  "code": "too_small",
+  "minimum": 1,
+  "type": "string",
+  "inclusive": true,
+  "exact": false,
+  "message": "Name cannot be blank",
+  "path": ["name"]
+}
+```
+
+Now, let's look at how we can implement this in our Next.js application. We'll start with the server action, then move on to the contact page, and finally the form component.
+
+## Server Action
+
+In our server action, we validate the form data using the Zod schema. If the validation fails, we return the array of errors. If the validation succeeds, we redirect the user to the home page.
+
+```ts
+// app/contact/server-action.ts
+"use server";
+
+import { redirect } from "next/navigation";
+import { z } from "zod";
+
+const schema = z.object({
+  name: z.string().min(1, "Name cannot be blank"),
+  email: z.string().email().min(1, "Email cannot be blank"),
+});
+
+export default async function contactAction(_prevState: any, params: FormData) {
+  const validation = schema.safeParse({
+    name: params.get("name"),
+    email: params.get("email"),
+  });
+
+  if (validation.success) {
+    // save the data, send an email, etc.
+    redirect("/");
+  } else {
+    return {
+      errors: validation.error.issues,
+    };
+  }
+}
+```
+
+## Contact Page
+
+In our contact page, we import the server action and pass it to the `Form` component. The `ContactPage` is a server component while the `Form` component is a client component. We need to use the `useFormState` hook to handle validation errors coming from the server. The `useFormState` hook can only be used in client components.
+
+```tsx
+// app/contact/page.tsx
+import contactAction from "./server-action";
+import Form from "./form";
+
+export default async function ContactPage() {
+  return (
+    <>
+      <h1 className="text-2xl font-bold mb-3">Contact Us</h1>
+      <Form action={contactAction}></Form>
+    </>
+  );
+}
+```
+
+## Form Component
+
+In our form component, we use the `useFormState` hook to handle validation errors. We match the `path` node of each error object to the corresponding input field and extract all `message` nodes to build the error message for each field.
+
+```tsx
+// app/contact/form.tsx
+"use client";
+
+import { useFormState, useFormStatus } from "react-dom";
+import type { ZodIssue } from "zod";
+
+type Props = {
+  action: (
+    _prevState: any,
+    params: FormData
+  ) => Promise<{ errors: ZodIssue[] }>;
+};
```

Comment:
As a reader, I'd like to know what this type definition is doing. Can you explain?

Reply:
Going to assume the reader is familiar with typescript already and do not want to tangent from the main concepts too much.